pkg/container/dagger_runner.go

```diff
@@ -0,0 +1,194 @@
+		client: d.client,
+		tmpDir: d.tmpDir,
+	}
+}
+
+// TempDir returns the base for temporary directory. For dagger, this is empty.
+func (d *daggerRunner) TempDir() string {
+	return ""
+}
+
+// StartPod starts a pod if necessary.  On Bubblewrap, we just run
+// ldconfig to prime ld.so.cache for glibc < 2.37 builds.
+func (d *daggerRunner) StartPod(ctx context.Context, cfg *Config) error {
+	_, span := otel.Tracer("melange").Start(ctx, "dagger.StartPod")
+	defer span.End()
+
+	// Initialize the Container from disk
+	imgPath := filepath.Join(d.tmpDir, imageTarName)
+	d.container = d.client.Container().Import(d.client.Host().File(imgPath))
+
+	// Add our cache dir
+	d.container = d.container.WithMountedCache("/var/cache/melange/", d.client.CacheVolume("build-cache"))
+
+	for _, mnt := range cfg.Mounts {
+
+		// We skip mounting in some files that we don't need in this mode
+		if mnt.Source == "/etc/resolv.conf" {
+			continue
+		}
+
+		// check if its a file or a directory
+		fi, err := os.Stat(mnt.Source)
+		if err != nil {
+			return err
+		}
+
+		d.logger.Infof("mounting %s to %s", mnt.Source, mnt.Destination)
+
+		if fi.IsDir() {
+			host := d.client.Host().Directory(mnt.Source)
+			d.container = d.container.WithMountedDirectory(mnt.Destination, host)
+		} else {
+			host := d.client.Host().File(mnt.Source)
+			d.container = d.container.WithMountedFile(mnt.Destination, host)
+		}
+	}
+
+	script := "[ -x /sbin/ldconfig ] && /sbin/ldconfig /lib || true"
+	d.container = d.container.WithExec([]string{"/bin/sh", "-c", script})
+	return nil
+}
+
+// TerminatePod terminates a pod if necessary.  Not implemented
+// for Bubblewrap runners.
+func (d *daggerRunner) TerminatePod(ctx context.Context, cfg *Config) error {
+	return nil
+}
+
+func (d *daggerRunner) WorkspaceTar(ctx context.Context, cfg *Config) (io.ReadCloser, error) {
+	return nil, nil
+}
+
+// WorkspaceTar implements Runner
+func (d *daggerRunner) Export(ctx context.Context, p string) error {
+	ctx, span := otel.Tracer("melange").Start(ctx, "docker.Export")
+	defer span.End()
+	output := d.container.Directory("/home/build/melange-out")
+	if _, err := output.Export(ctx, p+"/melange-out"); err != nil {
+		return err
+	}
+	return nil
+}
+
+type daggerLoader struct {
+	tmpDir string
+	client *dagger.Client
+}
+
+func (d *daggerLoader) LoadImage(ctx context.Context, layer v1.Layer, arch apko_types.Architecture, bc *apko_build.Context) (string, error) {
+	ctx, span := otel.Tracer("melange").Start(ctx, "docker.LoadImage")
+	defer span.End()
+
+	creationTime, err := bc.GetBuildDateEpoch()
+	if err != nil {
+		return "", err
+	}
+
+	img, err := apko_oci.BuildImageFromLayer(layer, bc.ImageConfiguration(), creationTime, arch, bc.Logger())
+	if err != nil {
+		return "", err
+	}
+
+	ref, err := name.ParseReference("melange:latest")
+	if err != nil {
+		return "", err
+	}
+
+	tarPath := filepath.Join(d.tmpDir, imageTarName)
+	if err := tarball.WriteToFile(tarPath, ref, img); err != nil {
```

Comment:
This will be really inefficient. I need to look to see if dagger exposes a better interface.

Reply:
Yea, I know. I think the only better option is to round trip through a registry. I looked and couldn't find much else.

Reply:
OR change the entire setup so dagger also invokes apko for us.